### PR TITLE
Specify level of TLS verification for database connections

### LINKED ISSUE #7636: Add ability to specify level of TLS verification for database connections
### What

Currently, Teleport database service always performs full TLS certificate verification when connecting to a target database. This is most secure but inflexible - some teams use their own certificates for their databases which do not always have proper hostname encoded, which they can't replace with Teleport-issued ones. For scenarios like this we need to add an option to lax verification of certificate presented by the database.

### How

Instead of introducing binary `insecure_skip_verify` flag, the proposal is to make level of TLS verification more flexible to support different customer use-cases (which have come up before).

Example database configuration:

```yaml
db_service:
  enabled: "yes"
  databases:
  - name: "example"
    protocol: "postgres"
    uri: "localhost:5432"
    ca_cert_file: xxx # deprecate this field and move it under "tls" section but keep for backwards compat.
    tls:
      mode: verify-full # can also be 'verify-ca' or 'insecure'
      server_name: PostgreSQL # optionally set custom server name
      ca_cert_file: /path/to/pem # optional CA path, existing field moved from the level above
```

* `verify-full` (default) does full verification like now. Server name can be optionally overridden, otherwise the one from URI is used.
* `verify-ca` skips hostname verification but checks CA.
* `insecure` is analogous to Go's "insecure_skip_verify" which accepts any certificate.

### PR DESCRIPTION
**Pull request by jakule**
This PR adds additional TLS section to the database configuration, for example:
```yaml
db_service:
  enabled: "yes"
  databases:
  - name: "example"
    protocol: "postgres"
    uri: "localhost:5432"
    ca_cert_file: xxx # deprecate this field and move it under "tls" section but keep for backwards compat.
    tls:
      mode: verify-full # can also be 'verify-ca' or 'insecure'
      server_name: PostgreSQL # optionally set custom server name
      ca_cert_file: /path/to/pem # optional CA path, existing field moved from the level above
```
`ca_cert_file` has been moved to `tls.ca_cert_file` (support for the old one has been preserved for backward compatibility).

closes #7636

### CODE REVIEW

**Comment by jakule** [comment]
I was able to pin down the failing tests. All the tests that are failing now, fail because of this error
```
failed to connect to `host=localhost user=postgres database=postgres`: dial error (dial tcp [::1]:41715: connect: cannot assign requested address)] db/server.go:639
```
This only happens when the Postgres client (https://github.com/gravitational/teleport/blob/4d5493eb6964b012dea8d0b1db7efc6980800287/lib/srv/db/postgres/engine.go#L214) tries to connect to the DB. `config.Fallbacks` by default resolves the host `localhost` to `127.0.0.1` and `::1` and then tries to connect to the DB. The first connection attempt returns a certificate validation error (expected from the test perspective). Then 
the driver tried to connect using IPv6, which fails as Docker by default doesn't enable it. This returns `cannot assign requested address` error, which is not the error expected by a test.

Here is the related postgres client code: https://github.com/gravitational/teleport/blob/efc99a068c883084902ae512feb3f1d9dbf4ca25/vendor/github.com/jackc/pgconn/pgconn.go#L116

Right now I'm not sure how can we fix that. From the test perspective, I'm not able to change that behavior (my only idea was to change the connection address from `localhost` to `127.0.0.1`, but this fails on certificate validation in other tests). I can skip the error message validation (this will make the test pass), but in the long run, should we make this configurable in any way? I'd expect some people to have the same problem.

**Comment by r0mant** [comment]
@jakule Nice RCA! I don't think skipping the error message is the best solution TBH. I have a couple of thoughts:

- We explicitly disable fallbacks to avoid retries here: https://github.com/gravitational/teleport/blob/v8.0.1/lib/srv/db/postgres/engine.go#L390. Does that still make the client retry with IPv6? Is there a way to disable it?
- Can we enable IPv6 in Docker?

**Comment by jakule** [comment]
@r0mant 
1. Yes, there is a way to disable IPv6. Look on this commit: https://github.com/gravitational/teleport/pull/9197/commits/7bee2f75c7abbc01ecb590bf16a1d1054f102701
Our CI is not complaining anymore :smile: 

2. Yes, but https://docs.docker.com/config/daemon/ipv6/
`Note: IPv6 networking is only supported on Docker daemons running on Linux hosts.`
so that test will be failing in many cases anyway. Also from what I see Google Cloud build doesn't support it.

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:390]: Looks like these will not have effect in case of cloud-hosted databases since we exit early above for them. I think the settings should apply to both self-hosted and cloud-hosted databases.

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:396]:
```suggestion
func VerifyConnectionFunc(rootCAs *x509.CertPool, serverName string) func(cs tls.ConnectionState) error {
```

Also, does this method need to be exported?

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:411]:
```suggestion
		return trace.Wrap(err)
```

File: lib/srv/db/postgres/engine.go
Comment (r0mant) [inline lib/srv/db/postgres/engine.go:391]: I wonder if we can disable ipv6 in the docker container where we're running tests?

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:381]: Nit: I think `switch dbTLSConfig.Mode` would work better here.

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:389]: Is there any difference between `verify-ca` and `verify-full` if ServerName is specified? Just wondering if we should be taking ServerName into account at all if the user requested `verify-ca` method.

File: lib/srv/db/common/auth.go
Comment (smallinsky) [inline lib/srv/db/common/auth.go:306]: The comment says "Don't set the ServerName when connecting to a MongoDB cluster" but if ServerName was provided the `tlsConfig.ServerName` will be set also for Mongo because first if condition `dbTLSConfig.ServerName != ""`  will be evaluated to true.

File: lib/config/configuration.go
Comment (smallinsky) [inline lib/config/configuration.go:1031]: Can we move this logic to a separate function ?

File: lib/srv/db/common/auth.go
Comment (r0mant) [reply lib/srv/db/common/auth.go:389]: Yeah I understand this is how it's implemented now. My question was rather, _should we_ take server name into account at all if the requested TLS verification method is "verify-ca"? I would think that server name should be irrelevant in that case since we're only going to validate that the certificate is signed by a proper authority.

Comment (r0mant) [reply lib/srv/db/common/auth.go:389]: Maybe we can also add a check to return configuration error if `server_name` is not empty when TLS mode is set to `verify-ca`.

File: api/types/database.go
Comment (jakule) [reply api/types/database.go:234]: @r0mant What should we do if the CA is provided anyway? I assumed that for cloud configuration that field will be empty and if someone decides to use it anyway at least will get a connection error. If we silently skip it, that can confuse some people.

File: lib/srv/db/common/auth.go
Comment (jakule) [reply lib/srv/db/common/auth.go:389]: @r0mant I like the idea of returning the error if `server_name` is not empty in `verify-ca` mode. I'd go even further in this case and also return an error if `server_name` is provided in `insecure` mode (basically only allow `server_name` only in `verify-full`).

File: api/types/database.go
Comment (jakule) [reply api/types/database.go:241]: Added

File: api/types/types.proto
Comment (jakule) [reply api/types/types.proto:328]: Added

File: lib/config/fileconf.go
Comment (jakule) [reply lib/config/fileconf.go:984]: Added

File: lib/service/cfg.go
Comment (jakule) [reply lib/service/cfg.go:640]: Added

File: lib/service/cfg.go
Comment (jakule) [reply lib/service/cfg.go:647]: Fixed

File: lib/service/cfg.go
Comment (jakule) [reply lib/service/cfg.go:654]: Added

File: lib/srv/db/common/auth.go
Comment (jakule) [reply lib/srv/db/common/auth.go:464]: Added

File: api/types/database.go
Comment (jakule) [reply api/types/database.go:241]: Added

File: api/types/types.proto
Comment (jakule) [reply api/types/types.proto]: Added

File: api/types/types.proto
Comment (jakule) [reply api/types/types.proto:328]: Added

File: api/types/types.proto
Comment (jakule) [reply api/types/types.proto]: I removed this enum and replace it with verify full with 0 value. Old habit: https://developers.google.com/protocol-buffers/docs/style

File: docs/pages/database-access/reference/configuration.mdx
Comment (jakule) [reply docs/pages/database-access/reference/configuration.mdx]: Fixed, thx

File: lib/config/configuration.go
Comment (jakule) [reply lib/config/configuration.go]: Fxied

File: lib/config/fileconf.go
Comment (jakule) [reply lib/config/fileconf.go]: Fixed

File: lib/config/fileconf.go
Comment (jakule) [reply lib/config/fileconf.go]: Fixed

File: lib/config/fileconf.go
Comment (jakule) [reply lib/config/fileconf.go]: Fixed

File: lib/srv/db/postgres/engine.go
Comment (jakule) [reply lib/srv/db/postgres/engine.go]: @r0mant I haven't been able to figure out this one. I checked many different things starting from changing the Docker configuration, checking the GCB options, overriding DNS resolver, listening on IPv6. I even tried to update the pgx to the latest version, but I wasn't able to make the tests pass in Docker. I have removed this callback, as I don't think that two tests are a good reason to remove IPv6 support for postgres. I also removed the error message check from the test to make them pass. If you have any other ideas let me know.

File: lib/config/configuration.go
Comment (jakule) [reply lib/config/configuration.go]: Sure

File: api/types/types.proto
Comment (jakule) [reply api/types/types.proto]: Updated

**Review by jakule — COMMENTED** [review]
@r0mant @smallinsky I addressed all comments and refactored the TLS logic.

File: lib/config/configuration.go
Comment (r0mant) [inline lib/config/configuration.go:1076]: Nit:
```suggestion
			return nil, trace.ConvertSystemError(err)
```

File: lib/config/configuration.go
Comment (r0mant) [inline lib/config/configuration.go:1085]: Let's include database name in the warning message.

File: lib/config/configuration.go
Comment (r0mant) [inline lib/config/configuration.go:1088]: Similar to above, let's include the database name to help identify problematic config section.

File: lib/config/configuration.go
Comment (r0mant) [inline lib/config/configuration.go:1092]: Nit:
```suggestion
				return nil, trace.ConvertSystemError(err)
```

File: lib/service/cfg.go
Comment (r0mant) [inline lib/service/cfg.go:632]:
```suggestion
	// VerifyCA checks the certificate, but skips the server name verification.
```

File: lib/service/cfg.go
Comment (r0mant) [inline lib/service/cfg.go:648]: Nit: Usually in cases like this, we'd make a slice with all valid modes (e.g. `AllTLSModes = []string{VerifyFull, VerifyCA, Insecure}`) and use it here instead of enumerating each individually.

File: lib/srv/db/postgres/engine.go
Comment (r0mant) [inline lib/srv/db/postgres/engine.go:391]: Nit: Need this change? :)

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:445]: This might be a good method to add to the actual `Database` resource interface e.g. `IsCloudHosted` or something like this (I actually thought about it too).

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:306]: This looks pretty clean :+1:

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:332]: Is there a reason to return half-built config?
```suggestion
		return nil, trace.Wrap(err)
```

File: lib/srv/db/common/auth.go
Comment (r0mant) [inline lib/srv/db/common/auth.go:401]:
```suggestion
		return nil, trace.Wrap(err)
```

**Review by r0mant — APPROVED** [review]
The logic lgtm now, I left a few stylistic/org comments - PTAL at them before you merge and wait for 2nd approval from @smallinsky.

Also, let's please make sure backward compatibility w/r/t `ca_cert_file` field (unless you have already):

- [ ] Connect new (built off of this branch) database agent that sets it in the new location with the old auth server.
- [ ] Connect old database agent that still has it in the old location with the new (off of this branch) auth server.

And please backport to `branch/v8`.

**Review by smallinsky — APPROVED** [review]

File: lib/config/configuration.go
Comment (jakule) [reply lib/config/configuration.go:1076]: Fixed

File: lib/config/configuration.go
Comment (jakule) [reply lib/config/configuration.go:1085]: Added

File: lib/config/configuration.go
Comment (jakule) [reply lib/config/configuration.go:1088]: Added

File: lib/config/configuration.go
Comment (jakule) [reply lib/config/configuration.go:1092]: Fixed

File: lib/service/cfg.go
Comment (jakule) [reply lib/service/cfg.go:632]: Fixed

File: lib/service/cfg.go
Comment (jakule) [reply lib/service/cfg.go:648]: Added

File: lib/srv/db/postgres/engine.go
Comment (jakule) [reply lib/srv/db/postgres/engine.go:391]: Nope, leftover after the custom IP resolved. Removed.

File: lib/srv/db/common/auth.go
Comment (jakule) [reply lib/srv/db/common/auth.go:445]: Moved.

File: lib/srv/db/common/auth.go
Comment (jakule) [reply lib/srv/db/common/auth.go:332]: Bad IDE refactor output. Fixed.

File: lib/srv/db/common/auth.go
Comment (jakule) [reply lib/srv/db/common/auth.go:401]: Fixed

**Comment by jakule** [comment]
@klizhentas Friendly ping. Can you look on the docs changes?

**Comment by jakule** [comment]
@klizhentas Please take a look at the docs changes. Thanks.

**Review by klizhentas — APPROVED** [review]

**Pull request merged**
